service/src/main/kotlin/app/cash/backfila/dashboard/CloneBackfillAction.kt

```diff
@@ -0,0 +1,217 @@
+import app.cash.backfila.service.persistence.DbRegisteredBackfill
+import app.cash.backfila.service.persistence.DbRunPartition
+import app.cash.backfila.service.persistence.DbService
+import misk.MiskCaller
+import misk.exceptions.BadRequestException
+import misk.hibernate.Id
+import misk.hibernate.Query
+import misk.hibernate.Transacter
+import misk.hibernate.load
+import misk.hibernate.loadOrNull
+import misk.logging.getLogger
+import misk.scope.ActionScoped
+import misk.security.authz.Authenticated
+import misk.web.PathParam
+import misk.web.Post
+import misk.web.RequestBody
+import misk.web.RequestContentType
+import misk.web.ResponseContentType
+import misk.web.actions.WebAction
+import misk.web.mediatype.MediaTypes
+import okio.ByteString
+import okio.ByteString.Companion.encodeUtf8
+import javax.inject.Inject
+
+enum class RangeCloneType {
+  RESTART,
+  CONTINUE,
+  NEW
+}
+
+data class CloneBackfillRequest(
+    // TODO move defaults to UI
+    val scan_size: Long = 1000,
+    val batch_size: Long = 100,
+    val num_threads: Int = 5,
+    val range_clone_type: RangeCloneType,
+    val pkey_range_start: String? = null,
+    val pkey_range_end: String? = null,
+    // Parameters that go to the client service.
+    val parameter_map: Map<String, ByteString> = mapOf(),
+    val dry_run: Boolean = true,
+    val backoff_schedule: String? = null,
+    // Sleep that is added after every successful RunBatch.
+    val extra_sleep_ms: Long = 0
+)
+
+data class CloneBackfillResponse(
+    val id: Long
+)
+
+class CloneBackfillAction @Inject constructor(
+    private val caller: @JvmSuppressWildcards ActionScoped<MiskCaller?>,
+    @BackfilaDb private val transacter: Transacter,
+    private val queryFactory: Query.Factory,
+    private val connectorProvider: ConnectorProvider
+) : WebAction {
+
+  @Post("/backfills/{id}/clone")
+  @RequestContentType(MediaTypes.APPLICATION_JSON)
+  @ResponseContentType(MediaTypes.APPLICATION_JSON)
+  // TODO allow any user
+  @Authenticated(capabilities = ["users"])
+  fun create(
+      @PathParam id: Long,
+      @RequestBody request: CloneBackfillRequest
+  ): CloneBackfillResponse {
+    // TODO check user has permissions for this service with access api
+
+    logger.info { "Clone backfill $id by ${caller.get()?.user}" }
+
+    if (request.num_threads < 1) {
+      throw BadRequestException("num_threads must be >= 1")
+    }
+    if (request.scan_size < 1) {
+      throw BadRequestException("scan_size must be >= 1")
+    }
+    if (request.batch_size < 1) {
+      throw BadRequestException("batch_size must be >= 1")
+    }
+    if (request.scan_size < request.batch_size) {
+      throw BadRequestException("scan_size must be >= batch_size")
+    }
+    if (request.extra_sleep_ms < 0) {
+      throw BadRequestException("extra_sleep_ms must be >= 0")
+    }
+    request.backoff_schedule?.let { schedule ->
+      if (schedule.split(',').any { it.toLongOrNull() == null }) {
```

Comment:
Perhaps there should be a more general place for these validations?

Reply:
in a followup I can probably refactor to have some common code between create/clone

Reply:
sounds good.